### PR TITLE
[peft] fix: Add replica_id handling for dense LoRA adapters with TP > 1

### PR DESCRIPTION
**Pull request by yaoyu-33**
## Summary

When using LoRA adapters on dense layers (non-MoE) with TP > 1, only one TP shard was being saved during checkpointing. This caused significant training-inference loss discrepancy because other TP ranks loaded zero/uninitialized adapter weights.

## Root Cause

In `ParallelLinearAdapter.sharded_state_dict()`, there was special `replica_id` handling for expert adapters (added in PR #1564, related to verl-project/verl#4303), but dense adapters never received the equivalent fix.

For dense adapters with TP > 1:
- Each TP rank generates a ShardedTensor with the same `replica_id`
- Shards should be distinguished by `global_offset`, but during the PEFT-filtered checkpoint save, TP shards are incorrectly deduplicated
- Result: Only TP rank 0's shard is saved

## Solution

Add `replica_id` handling for dense adapters similar to expert adapters. When TP > 1, the `replica_id` is adjusted to include the TP rank, ensuring each TP shard is correctly identified and saved during PEFT-filtered checkpoint saves.

## Changes
- Add `replica_id` adjustment for dense adapters with TP > 1 in `ParallelLinearAdapter.sharded_state_dict()`
- Add unit tests for the fix covering:
  - Dense adapters with TP > 1 (replica_id correctly updated)
  - Dense adapters with TP = 1 (no change to replica_id)
  - Expert adapters (still use EP-based replica_id, not affected)

## Testing
- [x] Unit tests added
- [ ] Functional tests with multi-GPU TP > 1 (requires CI)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed replica ID assignment for adapter checkpoints when using tensor model parallel configurations, ensuring correct deduplication during checkpoint saving.

* **Tests**
  * Added comprehensive test coverage for dense and expert adapter scenarios across different parallel configurations.

### CODE REVIEW

File: tests/functional_tests/training/test_finetune_lora.py [hunk 1]
```diff
@@ -443,3 +443,473 @@ def _setup_directories(self, base_dir, suffix=""):
             os.makedirs(lora_tensorboard_dir, exist_ok=True)
 
         return pretrain_checkpoint_dir, pretrain_tensorboard_dir, lora_checkpoint_dir, lora_tensorboard_dir
+
+
+class TestLoRAFinetuneTP2:
```
Comment (priyatham-resolve) [inline tests/functional_tests/training/test_finetune_lora.py:453]: nit: This seems to duplicate all the helper methods from `TestLoRAFinetune` above. Could share them via a base class or mixin to reduce the ~300 lines of boilerplate? Not a blocker.

**Review by priyatham-resolve — APPROVED** [review]
LGTM. Thanks for the fix

**Comment by yaoyu-33** [comment]
Cannot reproduce, defer to later.
